lib/bulk/common.js

```diff
@@ -427,26 +460,560 @@ function handleMongoWriteConcernError(batch, bulkResult, ordered, err, callback)
+    // TODO Bring from driver information in isMaster
+    // Get the namespace for the write operations
+    const namespace = collection.collectionName;
+    // Used to mark operation as executed
+    const executed = false;
+
+    // Current item
+    const currentOp = null;
+
+    // Handle to the bson serializer, used to calculate running sizes
+    const bson = topology.bson;
+
+    // Set max byte size
+    const maxBatchSizeBytes =
+      topology.isMasterDoc && topology.isMasterDoc.maxBsonObjectSize
+        ? topology.isMasterDoc.maxBsonObjectSize
+        : 1024 * 1025 * 16;
+    const maxWriteBatchSize =
+      topology.isMasterDoc && topology.isMasterDoc.maxWriteBatchSize
+        ? topology.isMasterDoc.maxWriteBatchSize
+        : 1000;
+
+    // Get the write concern
+    var writeConcern = applyWriteConcern(
+      shallowClone(options),
+      { collection: collection },
+      options
+    );
+    writeConcern = writeConcern.writeConcern;
+
+    // Get the promiseLibrary
+    const promiseLibrary = options.promiseLibrary || Promise;
+
+    // Final results
+    const bulkResult = {
+      ok: 1,
+      writeErrors: [],
+      writeConcernErrors: [],
+      insertedIds: [],
+      nInserted: 0,
+      nUpserted: 0,
+      nMatched: 0,
+      nModified: 0,
+      nRemoved: 0,
+      upserted: []
+    };
+
+    // Internal state
+    bulk.s = {
+      // Final result
+      bulkResult: bulkResult,
+      // Current batch state
+      currentBatch: null,
+      currentIndex: 0,
+      // ordered specific
+      currentBatchSize: 0,
+      currentBatchSizeBytes: 0,
+      // unordered specific
+      currentInsertBatch: null,
+      currentUpdateBatch: null,
+      currentRemoveBatch: null,
+      batches: [],
+      // Write concern
+      writeConcern: writeConcern,
+      // Max batch size options
+      maxBatchSizeBytes: maxBatchSizeBytes,
+      maxWriteBatchSize: maxWriteBatchSize,
+      // Namespace
+      namespace: namespace,
+      // BSON
+      bson: bson,
+      // Topology
+      topology: topology,
+      // Options
+      options: options,
+      // Current operation
+      currentOp: currentOp,
+      // Executed
+      executed: executed,
+      // Collection
+      collection: collection,
+      // Promise Library
+      promiseLibrary: promiseLibrary,
+      // Fundamental error
+      err: null,
+      // check keys
+      checkKeys: typeof options.checkKeys === 'boolean' ? options.checkKeys : true
+    };
+
+    // bypass Validation
+    if (options.bypassDocumentValidation === true) {
+      bulk.s.bypassDocumentValidation = true;
+    }
+  }
+
+  /**
+   * Add a single insert document to the bulk operation
+   *
+   * @param {object} document the document to insert
+   * @throws {MongoError}
+   * @return {OrderedBulkOperation or UnorderedBulkOperation}
+   */
+  insert(document) {
+    if (this.s.collection.s.db.options.forceServerObjectId !== true && document._id == null)
```

Comment:
what is `this.s`? It is not defined above in the constructor.

Reply:
it appears we refer to `this.s` a bunch in the base class. These _may_ be defined in subclasses, but if that's the case then the constructor for the base class should take the internal state as a parameter and extend it in the base. It should be _very_ clear by reading the code what is going on which is not the case right now.